### PR TITLE
Cleanup genericapiserver handler chain

### PR DESCRIPTION
**Pull request by sttts**
- move generic (api independent) handler filters to `pkg/genericapiserver/filters`
- entangle `genericapiserver.New()`
- unify signature of all handler filters (also those in `pkg/apiserver`)

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33164)

### CODE REVIEW

File: pkg/api/requestcontext.go [hunk 1]
```diff
@@ -89,21 +91,21 @@ func (c *requestContextMap) remove(req *http.Request) {
...
-		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
-			if mapper.init(req, NewContext()) {
-				// If we were the ones to successfully initialize, pair with a remove
-				defer mapper.remove(req)
-			}
-			handler.ServeHTTP(w, req)
-		}), nil
-	} else {
-		return handler, errors.New("Unknown RequestContextMapper implementation.")
+func WithRequestContext(handler http.Handler, mapper RequestContextMapper) http.Handler {
```
Comment (deads2k) [inline pkg/api/requestcontext.go]: Any idea why it's in this package?  Doesn't have to be this pull, but unless there's a strong reason, I'd like it moved eventually.

File: pkg/apiserver/handlers.go [hunk 2]
```diff
@@ -462,7 +144,7 @@ func (r *requestAttributeGetter) GetAttribs(req *http.Request) authorizer.Attrib
 }
 
 // WithAuthorizationCheck passes all authorized requests on to handler, and returns a forbidden error otherwise.
-func WithAuthorizationCheck(handler http.Handler, getAttribs RequestAttributeGetter, a authorizer.Authorizer) http.Handler {
+func WithAuthorization(handler http.Handler, getAttribs RequestAttributeGetter, a authorizer.Authorizer) http.Handler {
```
Comment (deads2k) [inline pkg/apiserver/handlers.go]: If a == nil, let's be friendly, skip adding this, and glog a warning.

File: pkg/genericapiserver/config.go [hunk 3]
```diff
@@ -337,18 +346,64 @@ func (c Config) New() (*GenericAPIServer, error) {
 		})
 	}
 
+	if len(c.AuditLogPath) != 0 {
```
Comment (deads2k) [inline pkg/genericapiserver/config.go]: Let's add to our running issue.  I think all the audit stuff should be gathered up into a pointer to a struct.

Comment (deads2k) [inline pkg/genericapiserver/config.go]: Not in this pull.

File: pkg/api/requestcontext.go (hunk 1 above)
Comment (sttts) [inline pkg/api/requestcontext.go]: Had the same question. Also the `Context` interface looks api independent. Doesn't have to be so deep in the package hierarchy. A good topic for a follow-up PR.

File: pkg/genericapiserver/config.go [hunk 4]
```diff
@@ -337,18 +346,64 @@ func (c Config) New() (*GenericAPIServer, error) {
...
-	// We do not register this using restful Webservice since we do not want to surface this in api docs.
-	// Allow GenericAPIServer to be embedded in contexts which already have something registered at the root
+	s.installAPI(&c)
+	s.Handler, s.InsecureHandler = s.buildHandlerChains(&c, http.Handler(s.Mux.BaseMux().(*http.ServeMux)))
+
+	return s, nil
+}
+
+func (s *GenericAPIServer) buildHandlerChains(c *Config, handler http.Handler) (secure http.Handler, insecure http.Handler) {
+	longRunningRE := regexp.MustCompile(c.LongRunningRequestRE)
```
Comment (deads2k) [inline pkg/genericapiserver/config.go]: Can't we push these into `WithTimeoutForNonLongRunningRequests`?

File: pkg/genericapiserver/config.go [hunk 5]
```diff
@@ -337,18 +346,64 @@ func (c Config) New() (*GenericAPIServer, error) {
...
+	longRunningFunc := genericfilters.BasicLongRunningRequestCheck(longRunningRE, map[string]string{"watch": "true"})
+
+	// common filters
+	handler = genericfilters.WithCORS(handler, allowedOriginRegexps(c.CorsAllowedOriginList), nil, nil, "true")
```
Comment (deads2k) [inline pkg/genericapiserver/config.go]: `common`?

File: pkg/apiserver/handlers.go (hunk 2 above)
Comment (sttts) [inline pkg/apiserver/handlers.go]: Most filters just return the passed handler when the parameters are nil. Will do the same here.

File: pkg/genericapiserver/config.go [hunk 6]
```diff
@@ -337,18 +346,64 @@ func (c Config) New() (*GenericAPIServer, error) {
...
+	// insecure filters
+	insecure = handler
+	insecure = api.WithRequestContext(insecure, c.RequestContextMapper)
+	insecure = genericfilters.WithTimeoutForNonLongRunningRequests(insecure, longRunningFunc)
+
+	// secure filters
+	attributeGetter := apiserver.NewRequestAttributeGetter(c.RequestContextMapper, s.NewRequestInfoResolver())
+	secure = handler
+	secure = apiserver.WithAuthorization(secure, attributeGetter, c.Authorizer)
+	secure = audit.WithAudit(secure, attributeGetter, s.auditWriter) // before impersonation to read original user
```
Comment (deads2k) [inline pkg/genericapiserver/config.go]: Isn't this after impersonation right now?

File: pkg/genericapiserver/config.go [hunk 7]
```diff
@@ -337,18 +346,64 @@ func (c Config) New() (*GenericAPIServer, error) {
...
-	// Register root handler.
-	// We do not register this using restful Webservice since we do not want to surface this in api docs.
-	// Allow GenericAPIServer to be embedded in contexts which already have something registered at the root
+	s.installAPI(&c)
+	s.Handler, s.InsecureHandler = s.buildHandlerChains(&c, http.Handler(s.Mux.BaseMux().(*http.ServeMux)))
+
+	return s, nil
+}
+
+func (s *GenericAPIServer) buildHandlerChains(c *Config, handler http.Handler) (secure http.Handler, insecure http.Handler) {
```
Comment (deads2k) [inline pkg/genericapiserver/config.go]: So much cleaner.  TODO to type the function and take it as an optional struct value in genericapiserver.

File: pkg/genericapiserver/config.go [hunk 8]
```diff
@@ -337,18 +346,64 @@ func (c Config) New() (*GenericAPIServer, error) {
...
+	secure = apiserver.WithImpersonation(secure, c.RequestContextMapper, c.Authorizer)
+	secure = authhandlers.WithAuthentication(secure, c.RequestContextMapper, c.Authenticator, authhandlers.Unauthorized(c.SupportsBasicAuth))
+	secure = api.WithRequestContext(secure, c.RequestContextMapper)
+	secure = genericfilters.WithTimeoutForNonLongRunningRequests(secure, longRunningFunc)
+	secure = genericfilters.WithMaxInFlightLimit(secure, c.MaxRequestsInFlight, longRunningFunc)
+
+	return
+}
+
+func allowedOriginRegexps(allowedOrigins []string) []*regexp.Regexp {
```
Comment (deads2k) [inline pkg/genericapiserver/config.go]: Seems like this should live in your cors file

File: pkg/genericapiserver/filters/cors.go [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+package filters
+
+import (
+	"net/http"
+	"regexp"
+	"strings"
+)
+
+// TODO: use restful.CrossOriginResourceSharing
```
Comment (deads2k) [inline pkg/genericapiserver/filters/cors.go]: I don't think we're likely to do this.

File: pkg/genericapiserver/filters/cors.go [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+// See github.com/emicklei/go-restful/blob/master/examples/restful-CORS-filter.go, and
+// github.com/emicklei/go-restful/blob/master/examples/restful-basic-authentication.go
+// Or, for a more detailed implementation use https://github.com/martini-contrib/cors
+// or implement CORS at your proxy layer.
+
+// WithCORS is a simple CORS implementation that wraps an http Handler.
+// Pass nil for allowedMethods and allowedHeaders to use the defaults. If allowedOriginPatterns
+// is empty or nil, no CORS support is installed.
+func WithCORS(handler http.Handler, allowedOriginPatterns []*regexp.Regexp, allowedMethods []string, allowedHeaders []string, allowCredentials string) http.Handler {
```
Comment (deads2k) [inline pkg/genericapiserver/filters/cors.go]: straight move?

File: pkg/genericapiserver/genericapiserver.go [hunk 11]
```diff
@@ -165,14 +162,17 @@ type GenericAPIServer struct {
 	enableOpenAPISupport   bool
 	openAPIInfo            spec.Info
 	openAPIDefaultResponse spec.Response
+	openAPIDefinitions     *common.OpenAPIDefinitions
```
Comment (deads2k) [inline pkg/genericapiserver/genericapiserver.go]: didn't see this coming.  unrelated to this pull?

Comment (deads2k) [inline pkg/genericapiserver/genericapiserver.go]: oh, a move, nm

File: pkg/api/requestcontext.go (hunk 1 above)
Comment (deads2k) [inline pkg/api/requestcontext.go]: > Had the same question. Also the Context interface looks api independent. Doesn't have to be so deep in the package hierarchy. A good topic for a follow-up PR.

agreed.

File: pkg/genericapiserver/config.go (hunk 4 above)
Comment (sttts) [inline pkg/genericapiserver/config.go]: yes, makes sense

**Comment by deads2k** [comment]
Assuming most of the code is straight moves, I've got a few minor comments, but this looks a lot better.

Any specific reason its still WIP?  I'd be fine tying off this chunk without adding more to it.

File: pkg/genericapiserver/config.go (hunk 6 above)
Comment (sttts) [inline pkg/genericapiserver/config.go]: Good spot. In origin it's correct.

File: pkg/genericapiserver/config.go (hunk 8 above)
Comment (sttts) [inline pkg/genericapiserver/config.go]: moved

File: pkg/genericapiserver/filters/cors.go (hunk 9 above)
Comment (sttts) [inline pkg/genericapiserver/filters/cors.go]: Somebody somewhen was planning to do that :)

File: pkg/genericapiserver/filters/cors.go (hunk 10 above)
Comment (sttts) [inline pkg/genericapiserver/filters/cors.go]: yes

**Comment by deads2k** [comment]
lgtm.  squash and tag when you're happy with it.

File: pkg/genericapiserver/config.go [hunk 12]
```diff
@@ -382,8 +382,8 @@ func (s *GenericAPIServer) buildHandlerChains(c *Config, handler http.Handler) (
 	attributeGetter := apiserver.NewRequestAttributeGetter(c.RequestContextMapper, s.NewRequestInfoResolver())
 	secure = handler
 	secure = apiserver.WithAuthorization(secure, attributeGetter, c.Authorizer)
-	secure = audit.WithAudit(secure, attributeGetter, s.auditWriter) // before impersonation to read original user
 	secure = apiserver.WithImpersonation(secure, c.RequestContextMapper, c.Authorizer)
+	secure = audit.WithAudit(secure, attributeGetter, s.auditWriter) // before impersonation to read original user
```
Comment (soltysh) [inline pkg/genericapiserver/config.go]: :+1:

File: pkg/genericapiserver/config.go (hunk 5 above)
Comment (sttts) [inline pkg/genericapiserver/config.go]: common for secure and insecure.

Comment (deads2k) [inline pkg/genericapiserver/config.go]: > common for secure and insecure.

Yeah, I was suggesting a rename.  Doesn't really matter to me.

**Comment by sttts** [comment]
@k8s-bot gce e2e test this issue: #32574

**Comment by sttts** [comment]
@k8s-bot kubemark gce e2e test this issue: #32574

**Comment by sttts** [comment]
@k8s-bot test this issue: #32574

**Comment by sttts** [comment]
Squashed. Waiting for #33095 to merge.

**Comment by sttts** [comment]
Rebased.

**Pull request merged**